File: sdk/javascript/src/responses.js
export function dataSourceResponse(value) {
  const envelope = exactObject(value, ["source"], "data-source response");
  const source = exactObject(envelope.source, [
    "id", "name", "driver", "execution_mode", "host", "port", "database_name", "username", "ssl_mode",
    "use_dsn", "authority_url", "credential_env", "read_only", "password_set", "password_hint",
    "last_test_status", "last_test_message", "last_test_at", "catalog_updated_at", "created_at", "updated_at",
  ], "data source", true);
  if (!nonblank(source.id) || source.driver !== "postgres" || source.read_only !== true || !["direct", "delegated"].includes(source.execution_mode)) {
    throw new TypeError("Omnidex returned an invalid data-source authority.");
  }
  return source;
}

export function channelResponse(value, expectedId, expectedSourceId) {
  const envelope = exactObject(value, ["channel"], "channel response");
  const channel = validateChannel(envelope.channel);
  if (channel.id !== expectedId || (expectedSourceId !== undefined && channel.data_source_id !== expectedSourceId) ||
    channel.mode !== "assistant") {
    throw new TypeError("Omnidex returned a channel outside the requested authority.");
  }
  return channel;
}

export function messageResponse(value, channelId, prompt) {
  const envelope = exactObject(value, ["channel", "user_message", "job"], "message response");
  const channel = validateChannel(envelope.channel);
  const message = validateMessage(envelope.user_message);
  const job = validateJob(envelope.job);
  if (channel.id !== channelId || message.channel_id !== channelId || message.content !== prompt || job.id < 1) {
    throw new TypeError("Omnidex returned a message outside the requested authority.");
  }
  return { channel, user_message: message, job };
}

export function messagePageResponse(value, channelId) {
  const page = exactObject(value, ["channel_id", "messages", "next_before_id", "has_more"], "message page", true);
  if (page.channel_id !== channelId || !Array.isArray(page.messages) || typeof page.has_more !== "boolean") {
    throw new TypeError("Omnidex returned an invalid message-page authority.");
  }
  if (page.has_more !== (Number.isSafeInteger(page.next_before_id) && page.next_before_id > 0)) {
    throw new TypeError("Omnidex returned contradictory message pagination.");
  }
  return { ...page, messages: page.messages.map(validateMessage) };
}

export function jobDetailsResponse(value, jobId) {
  const details = exactObject(value, ["job", "steps", "contexts"], "job details");
  const job = validateJob(details.job);
  if (job.id !== jobId || !Array.isArray(details.steps) || !Array.isArray(details.contexts)) {
    throw new TypeError("Omnidex returned a different job authority.");
  }
  return {
    job,
    steps: details.steps.map((step) => exactObject(step, [
      "id", "job_id", "action", "sort_index", "status", "generation", "superseded_at_generation",
      "worker_id", "output", "error", "started_at", "finished_at", "created_at", "updated_at",
    ], "job step", true)),
    contexts: details.contexts.map((context) => exactObject(context, ["id", "step_id", "key", "value", "created_at"], "job context")),
  };
}

export function apiError(status, value) {
  let message = "invalid error envelope";
  try {
    const error = exactObject(value, ["error"], "error envelope");
    if (nonblank(error.error)) message = error.error;
  } catch {
    // The fixed message deliberately avoids reflecting an unvalidated response.
  }
  return new OmnidexApiError(status, message);
}

export class OmnidexApiError extends Error {
  constructor(status, message) {
    super(`Omnidex integration API failed with HTTP ${status}: ${message}`);
    this.name = "OmnidexApiError";
    this.status = status;
    this.apiMessage = message;
  }
}

function validateChannel(value) {
  return exactObject(value, [
    "id", "scope", "name", "tags", "project_id", "workspace_root", "data_source_id", "mode",
    "roleplay_viewpoint_character_id", "created_at", "updated_at",
  ], "channel", true);
}

function validateMessage(value) {
  const message = exactObject(value, ["id", "channel_id", "role", "content", "created_at"], "channel message");
  if (!Number.isSafeInteger(message.id) || message.id < 1 || !["user", "assistant"].includes(message.role) || !nonblank(message.channel_id)) {
    throw new TypeError("Omnidex returned an invalid channel message.");
  }
  return message;
}

function validateJob(value) {
  const job = exactObject(value, [
    "id", "instruction", "pipeline", "status", "result", "error", "metadata", "current_generation",
    "created_at", "updated_at", "completed_at",
  ], "job", true);
  if (!Number.isSafeInteger(job.id) || job.id < 1) throw new TypeError("Omnidex returned an invalid job identity.");
  return job;
}

function exactObject(value, allowed, label, optional = false) {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new TypeError(`${label} must be an object.`);
  const allowedSet = new Set(allowed);
  for (const key of Object.keys(value)) {
    if (!allowedSet.has(key)) throw new TypeError(`${label} contains unknown field ${JSON.stringify(key)}.`);
  }
  if (!optional) {
    for (const key of allowed) if (!(key in value)) throw new TypeError(`${label} is missing field ${JSON.stringify(key)}.`);
  }
  return value;
}

function nonblank(value) {
  return typeof value === "string" && value.trim().length > 0;
}

File: sdk/rust/src/responses.rs
use serde::de::DeserializeOwned;
use serde::Deserialize;

use crate::{
    Channel, ChannelMessage, DataSource, Error, JobDetails, MessagePage, SendMessageResult,
};

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct DataSourceEnvelope {
    source: DataSource,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct ChannelEnvelope {
    channel: Channel,
}

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct ErrorEnvelope {
    error: String,
}

pub(crate) fn data_source(raw: &[u8]) -> Result<DataSource, Error> {
    let source = decode::<DataSourceEnvelope>(raw)?.source;
    if source.id.trim().is_empty()
        || source.name.trim().is_empty()
        || source.driver != "postgres"
        || !matches!(source.execution_mode.as_str(), "direct" | "delegated")
        || !source.read_only
    {
        return protocol("Omnidex returned an invalid data-source authority.");
    }
    Ok(source)
}

pub(crate) fn channel(raw: &[u8], id: &str, source_id: Option<&str>) -> Result<Channel, Error> {
    let channel = decode::<ChannelEnvelope>(raw)?.channel;
    validate_channel(&channel)?;
    if channel.id != id
        || source_id.is_some_and(|expected| channel.data_source_id != expected)
        || channel.mode != "assistant"
    {
        return protocol("Omnidex returned a channel outside the requested authority.");
    }
    Ok(channel)
}

pub(crate) fn message(
    raw: &[u8],
    channel_id: &str,
    prompt: &str,
) -> Result<SendMessageResult, Error> {
    let result = decode::<SendMessageResult>(raw)?;
    validate_channel(&result.channel)?;
    validate_message(&result.user_message)?;
    validate_job(result.job.id)?;
    if result.channel.id != channel_id
        || result.user_message.channel_id != channel_id
        || result.user_message.content != prompt
    {
        return protocol("Omnidex returned a message outside the requested authority.");
    }
    Ok(result)
}

pub(crate) fn message_page(raw: &[u8], channel_id: &str) -> Result<MessagePage, Error> {
    let page = decode::<MessagePage>(raw)?;
    if page.channel_id != channel_id
        || page.has_more != page.next_before_id.is_some()
        || page.next_before_id.is_some_and(|id| id < 1)
    {
        return protocol("Omnidex returned contradictory message-page authority.");
    }
    for message in &page.messages {
        validate_message(message)?;
    }
    Ok(page)
}

pub(crate) fn job_details(raw: &[u8], job_id: i64) -> Result<JobDetails, Error> {
    let details = decode::<JobDetails>(raw)?;
    validate_job(details.job.id)?;
    if details.job.id != job_id {
        return protocol("Omnidex returned a different job authority.");
    }
    Ok(details)
}

pub(crate) fn api_error(status: u16, raw: &[u8]) -> Error {
    let message = decode::<ErrorEnvelope>(raw)
        .ok()
        .map(|value| value.error)
        .filter(|value| !value.trim().is_empty())
        .unwrap_or_else(|| "invalid error envelope".into());
    Error::Api { status, message }
}

fn validate_channel(channel: &Channel) -> Result<(), Error> {
    if channel.id.trim().is_empty()
        || channel.scope.trim().is_empty()
        || channel.mode.trim().is_empty()
    {
        return protocol("Omnidex returned an invalid channel authority.");
    }
    Ok(())
}

fn validate_message(message: &ChannelMessage) -> Result<(), Error> {
    if message.id < 1
        || message.channel_id.trim().is_empty()
        || !matches!(message.role.as_str(), "user" | "assistant")
    {
        return protocol("Omnidex returned an invalid channel message.");
    }
    Ok(())
}

fn validate_job(id: i64) -> Result<(), Error> {
    if id < 1 {
        return protocol("Omnidex returned an invalid job identity.");
    }
    Ok(())
}

fn decode<T: DeserializeOwned>(raw: &[u8]) -> Result<T, Error> {
    serde_json::from_slice(raw)
        .map_err(|error| Error::Protocol(format!("decode Omnidex response: {error}")))
}

fn protocol<T>(message: &str) -> Result<T, Error> {
    Err(Error::Protocol(message.into()))
}
